src/sentry/profiles/task.py

```diff
@@ -273,37 +307,67 @@ def _normalize(profile: Profile, organization: Organization) -> None:
             profile["device_classification"] = classification
 
 
-def _prepare_frames_from_profile(profile: Profile) -> Tuple[List[Any], List[Any], set[int]]:
+def _prepare_frames_from_profile(
+    profile: Profile, platform: Optional[str] = None
+) -> Tuple[List[Any], List[Any], set[int]]:
     with sentry_sdk.start_span(op="task.profiling.symbolicate.prepare_frames"):
         modules = profile["debug_meta"]["images"]
         frames: List[Any] = []
         frames_sent: set[int] = set()
 
+        if platform is None:
+            platform = profile["platform"]
+
         # NOTE: the usage of `adjust_instruction_addr` assumes that all
         # the profilers on all the platforms are walking stacks right from a
         # suspended threads cpu context
 
         # in the sample format, we have a frames key containing all the frames
         if "version" in profile:
-            if profile["platform"] in JS_PLATFORMS:
+            if platform in JS_PLATFORMS:
                 for idx, f in enumerate(profile["profile"]["frames"]):
                     if is_valid_javascript_frame(f, profile):
                         frames_sent.add(idx)
 
                 frames = [profile["profile"]["frames"][idx] for idx in frames_sent]
             else:
-                frames = profile["profile"]["frames"]
+                # if the root platform is cocoa, then we know we have only cocoa frames
+                if profile["platform"] == "cocoa":
+                    frames = profile["profile"]["frames"]
+                else:
+                    # else, we might have both js and cocoa ones (ract native)
+                    # and we need to filter only for the cocoa ones
+                    for idx, f in enumerate(profile["profile"]["frames"]):
+                        if (
+                            f.get("platform", "") == "cocoa"
+                            and f.get("instruction_addr") is not None
+                        ):
+                            frames_sent.add(idx)
+                    frames = [profile["profile"]["frames"][idx] for idx in frames_sent]
```

Comment:
The logic here seems a little strange. We're special casing `cocoa` here when the special case should be react native. Instead of this current flow, I think it makes more sense to have something like
```python
elif profile["platform"] != platform:  # indication that we're looking for native frames only like on react native
    for idx, f in enumerate(profile["profile"]["frames"]):
        if f.get("platform", "") == platform and f.get("instruction_addr") is not None:
            frames_sent.add(idx)
        frames = [profile["profile"]["frames"][idx] for idx in frames_sent]
else:
    frames = profile["profile"]["frames"]
```